amf-webapi/shared/src/main/scala/amf/plugins/document/webapi/parser/spec/jsonschema/ReferenceResolver.scala

```diff
@@ -0,0 +1,45 @@
+package amf.plugins.document.webapi.parser.spec.jsonschema
+
+import amf.plugins.document.webapi.parser.spec.common.YMapEntryLike
+import org.yaml.model.YMap
+import amf.core.parser.YMapOps
+import scala.util.Try
+import java.net.URI
+import scala.annotation.tailrec
+
+trait ReferenceResolver {
+  def resolve(reference: String, index: Map[String, YMapEntryLike]): Option[YMapEntryLike]
+}
+
+object FragmentTraversingResolver extends ReferenceResolver {
+
+  override def resolve(reference: String, index: Map[String, YMapEntryLike]): Option[YMapEntryLike] =
+    Try(new URI(reference)).toOption.flatMap { uri =>
+      val fragment = uri.getRawFragment
+      val baseUri  = reference.stripSuffix(s"#$fragment")
+      val baseNode = index.get(baseUri)
+      baseNode.flatMap { node =>
+        val keys = keysOfFragment(fragment)
+        keys.flatMap(traverseFragmentKeys(node, _))
+      }
+    }
+
+  @tailrec
+  private def traverseFragmentKeys(entry: YMapEntryLike, keys: List[String]): Option[YMapEntryLike] =
+    keys.headOption match {
+      case Some(key) =>
+        val currentMap = entry.value.asOption[YMap]
+        val newNode    = currentMap.flatMap(_.key(key))
+        newNode match {
+          case Some(entry) => traverseFragmentKeys(YMapEntryLike(entry), keys.tail)
+          case None        => None
+        }
+      case None => Some(entry)
+    }
+
+  private def keysOfFragment(fragment: String): Option[List[String]] = {
+    if (fragment.isEmpty) Some(Nil)
+    else if (fragment.startsWith("/")) Some(fragment.stripPrefix("/").split("/").toList)
```

Comment:
We should centralize this opinionated entry storage mechanism in the AstIndex or somewhere